Remove support for resolvelib < 0.8.0, since 0.8.0 is over 3 years old

s-hertel opened the pull request:
##### SUMMARY

Follow up to #84218.

@mattclay suggested some of the old 0.x versions should be dropped since 0.8.0 is several years old already, and this lets us simplify the code and reduce test time.

##### ISSUE TYPE

- Feature Pull Request

s-hertel committed:
Remove support for resolvelib < 0.8.0, since 0.8.0 is over 3 years old

Simplify the code and drop some versions from the test to reduce the
time it takes to run

s-hertel committed:
missed a spot

mattclay commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -38,16 +38,16 @@ class AbstractProvider:  # type: ignore[no-redef]
 
     def __init__(
-            self,  # type: CollectionDependencyProviderBase
+            self,  # type: CollectionDependencyProvider
```
Since this line is already being updated it would be nice to switch it to a native type hint.

webknjaz replied:
That would be just `typing.Self` or could be kept implicit.

mattclay replied:
Ah, good point. I was so focused on the type comment I didn't even realize it shouldn't be needed at all.

s-hertel replied:
Fixed by making it implicit.

webknjaz commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -103,47 +103,21 @@ def identify(self, requirement_or_candidate):
+        resolutions: t.Mapping[str, Candidate],
+        candidates: t.Mapping[str, t.Iterator[Candidate]],
+        information: t.Iterator[t.NamedTuple],
+        backtrack_causes: t.Sequence
```
`typing.Sequence` has been deprecated in Python 3.9: https://docs.python.org/3.14/library/typing.html#typing.Sequence. We should really use `collections.abc.Sequence` instead.

Also, it's a generic so to fully cover typing, an item type needs to be specified as well.

s-hertel replied:
fixed

webknjaz commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -103,47 +103,21 @@ def identify(self, requirement_or_candidate):
+        candidates: t.Mapping[str, t.Iterator[Candidate]],
+        information: t.Iterator[t.NamedTuple],
+        backtrack_causes: t.Sequence
+    ) -> t.Union[float, int]:
```
Can we use slightly newer syntax here?
```suggestion
    ) -> float | int:
```

s-hertel replied:
done

s-hertel committed:
Fix issues masked by the non-native type hints related to these changes

Address review comments

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -84,7 +82,7 @@ def _get_collection_versions(self, requirement: Requirement) -> t.Iterator[tuple
             raise last_error
 
-    def get_collection_versions(self, requirement: Requirement) -> t.Iterable[tuple[str, GalaxyAPI]]:
+    def get_collection_versions(self, requirement: Candidate) -> t.Iterable[tuple[str, GalaxyAPI]]:
```
This doesn't seem right. The method is called with an abstract requirement spec and not a concrete resolved candidate, correct?

s-hertel replied:
I don't disagree. I see a variety of issues with the ansible-galaxy type hints, but I've tried to limit what I'm changing in this PR. I can revisit whether we want to actually use fix the Candidates to make them Requirements in a followup, but I think now it at least reflects reality.

s-hertel replied:
I updated it to Requirement | Candidate to resolve the type hint failures with only Requirement in the interim.

webknjaz commented on test/integration/targets/ansible-galaxy-collection/vars/main.yml:
```diff
@@ -5,14 +5,10 @@ gpg_homedir: "{{ galaxy_dir }}/gpg"
-  - "0.7.0"  # test CollectionDependencyProvider070
-  - "<2.0.0"  # test CollectionDependencyProvider080
+  - "0.8.0"
+  - "<2.0.0"
```
*(nit)*
```suggestion
  - "< 2.0.0"
```

webknjaz commented:
Could you explain the changes from `Candidate` to `Requirement` in many places? Is there any typing difficulty it was causing? If so, please post the MyPy output and I'll be happy to take a look..

s-hertel committed:
Update test/integration/targets/ansible-galaxy-collection/vars/main.yml

Co-authored-by: 🇺🇦 Sviatoslav Sydorenko (Святослав Сидоренко) <wk.cvs.github@sydorenko.org.ua>

s-hertel commented:
> Could you explain the changes from Candidate to Requirement in many places? Is there any typing difficulty it was causing? If so, please post the MyPy output and I'll be happy to take a look..

Yep, I had the same thought. I have another branch where I've changed the type, but left the code alone for this change and just tried to make the type hints match reality.

s-hertel committed:
Change it to Candidate | Requirement to fix the ansible-test type hint errors when it's only Requirement

webknjaz commented:
I don't see where candidates are being passed into these functions. Trying to look deeper. But perhaps we could exclude the type updates from the PR and do this in a separate one?

webknjaz commented:
@s-hertel finally got a minute to look into it. I'll push the fix to your branch.

webknjaz committed:
Set `get_collection_versions` arg to `Requirement`

This method is only called with requirement instances, never with
candidates.

webknjaz committed:
Set `Requirement` type @ `find_matches()` arg

This method always gets `Requirement` instances in the corresponding
argument. Never candidates.

webknjaz committed:
Set `Requirement` type @ `is_satisfied_by()`

This method accepts a pair of a requirement and a candidate. Never two
candidates.

webknjaz committed:
Note `get_dependencies()` returns list of reqs

Previously, it incorrectly marked the return annotation as a list of
candidates.

webknjaz committed:
Revert the candidate/requirement formatting change

webknjaz commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -425,8 +390,7 @@ def is_satisfied_by(self, requirement, candidate):
 
-    def get_dependencies(self, candidate):
-        # type: (Candidate) -> list[Candidate]
+    def get_dependencies(self, candidate: Candidate) -> list[Requirement]:
```
*(@s-hertel I think this return value was the problem that confused you)*

webknjaz commented on test/integration/targets/ansible-galaxy-collection-scm/vars/main.yml:
```diff
@@ -5,7 +5,5 @@ test_repo_path: "{{ galaxy_dir }}/development/ansible_test"
-  - "0.7.0"
-  - "0.8.0"
+  - "0.8.0"  # Oldest supported
+  - "<2.0.0"
```
```suggestion
  - "< 2.0.0"
```

webknjaz committed:
Make `< 2.0.0` in the version spec sparse

webknjaz commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -103,47 +105,21 @@ def identify(self, requirement_or_candidate):
+        identifier: str,
+        resolutions: t.Mapping[str, Candidate],
+        candidates: t.Mapping[str, t.Iterator[Candidate]],
+        information: t.Iterator[t.NamedTuple],
```
Looking at resolvelib, this could be more precise (imprecise types contribute to incomplete coverage). But that's something to tackle separately.

webknjaz commented:
/azp run

mattclay commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -103,47 +105,21 @@ def identify(self, requirement_or_candidate):
+    def get_preference(
+        self,
+        identifier: str,
+        resolutions: t.Mapping[str, Candidate],
```
Abstract container types like `Mapping` and `Iterator` in `collections.abc` have supported subscripting since Python 3.9. We should be using those in controller code instead of the deprecated ones in the `typing` module. In ansible-core 2.20 we'll be able to use them for target code as well.

mattclay commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -5,6 +5,8 @@
 
 from __future__ import annotations
 
+from collections.abc import Sequence
```
Since we often end up using many `collections.abc` imports for type hints, a common pattern we use is to import the entire module and alias it like we do the `typing` module:

```python
import collections.abc as c
```

You can use `_c` instead if you're taking care to avoid including imports in the public interface of a module.

mattclay reviewed (approved):
The changes look good.

If you want to clean up a few of the type hints, see my other comments.

mattclay commented:
/azp run

s-hertel commented:
I've reopened this as https://github.com/ansible/ansible/pull/85936.
